src/codegen/encoding/borsh_encoding.rs

```diff
@@ -0,0 +1,624 @@
+                let get_size = Expression::Builtin(
+                    Loc::Codegen,
+                    vec![Type::Uint(32)],
+                    Builtin::ArrayLength,
+                    vec![expr.clone()],
+                );
+                let array_length = vartab.temp_anonymous(&Type::Uint(32));
+                cfg.add(
+                    vartab,
+                    Instr::Set {
+                        loc: Loc::Codegen,
+                        res: array_length,
+                        expr: get_size,
+                    },
+                );
+
+                let var = Expression::Variable(Loc::Codegen, Type::Uint(32), array_length);
+                cfg.add(
+                    vartab,
+                    Instr::WriteBuffer {
+                        buf: buffer.clone(),
+                        offset: offset.clone(),
+                        value: var.clone(),
+                    },
+                );
+
+                // ptr + offset + size_of_integer
+                let dest_address = Expression::AdvancePointer {
+                    loc: Loc::Codegen,
+                    ty: Type::BufferPointer,
+                    pointer: Box::new(buffer.clone()),
+                    bytes_offset: Box::new(increment_four(offset.clone())),
+                };
+
+                cfg.add(
+                    vartab,
+                    Instr::MemCopy {
+                        source: expr.clone(),
+                        destination: dest_address,
+                        bytes: var.clone(),
+                    },
+                );
+
+                increment_four(var)
+            }
+
+            Type::Enum(_) => {
+                cfg.add(
+                    vartab,
+                    Instr::WriteBuffer {
+                        buf: buffer.clone(),
+                        offset: offset.clone(),
+                        value: expr.clone(),
+                    },
+                );
+
+                Expression::NumberLiteral(Loc::Codegen, Type::Uint(32), BigInt::one())
+            }
+
+            Type::Struct(struct_no) => self.encode_struct(
+                expr,
+                buffer,
+                offset.clone(),
+                &expr_ty,
+                *struct_no,
+                arg_no,
+                ns,
+                vartab,
+                cfg,
+            ),
+
+            Type::Array(ty, dims) => self.encode_array(
+                expr, &expr_ty, ty, dims, arg_no, buffer, offset, ns, vartab, cfg,
+            ),
+
+            Type::UserType(_) | Type::Unresolved | Type::Rational | Type::Unreachable => {
+                unreachable!("Type should not exist in codegen")
+            }
+
+            Type::InternalFunction { .. }
+            | Type::ExternalFunction { .. }
+            | Type::Void
+            | Type::BufferPointer
+            | Type::Mapping(..) => unreachable!("This type cannot be encoded"),
```

Comment:
`Type::ExternalFunction` was not support by eth abi encoding because it would be incompatible.

With borsh encoding we can encode this (4 bytes function selector + address)